devito/petsc/iet/routines.py

```diff
@@ -383,33 +418,263 @@ def create_formrhs_body(self, injectsolve, body, solver_objs, objs):
 
         # Replace non-function data with pointer to data in struct
         subs = {i._C_symbol: FieldFromPointer(i._C_symbol, struct) for
-                i in struct.fields if not isinstance(i.function, AbstractFunction)}
-
+                i in fields if not isinstance(i.function, AbstractFunction)}
         formrhs_body = Uxreplace(subs).visit(formrhs_body)
 
-        self._struct_params.extend(struct.fields)
+        self._struct_params.extend(fields)
 
         return formrhs_body
 
-    def runsolve(self, solver_objs, objs, rhs_callback, injectsolve):
+    def local_struct(self, solver_objs):
+        """
+        This is the struct used within callback functions,
+        usually accessed via DMGetApplicationContext.
+        """
+        return petsc_struct(
+            solver_objs['dummyctx'].name,
+            self.filtered_struct_params,
+            solver_objs['Jac'].name+'_ctx',
+            liveness='eager'
+        )
+
+    def main_struct(self, solver_objs):
+        """
+        This is the struct initialised inside the main kernel and
+        attached to the DM via DMSetApplicationContext.
+        """
+        return petsc_struct(
+            self.sregistry.make_name(prefix='ctx'),
+            self.filtered_struct_params,
+            solver_objs['Jac'].name+'_ctx'
+        )
+
+    def make_struct_callback(self, solver_objs, objs):
+        mainctx = solver_objs['mainctx']
+        body = [
+            DummyExpr(FieldFromPointer(i._C_symbol, mainctx), i._C_symbol)
+            for i in mainctx.callback_fields
+        ]
+        struct_callback_body = CallableBody(
+            List(body=body), init=(petsc_func_begin_user,),
+            retstmt=tuple([Call('PetscFunctionReturn', arguments=[0])])
+        )
+        struct_callback = Callable(
+            self.sregistry.make_name(prefix='PopulateMatContext_'),
+            struct_callback_body, objs['err'],
+            parameters=[mainctx]
+        )
+        self._efuncs[struct_callback.name] = struct_callback
+        self._struct_callback = struct_callback
+
+    def dummy_fields(self, iet, solver_objs):
+        # Place all context data required by the shell routines into a struct
+        fields = [f.function for f in FindSymbols('basics').visit(iet)]
+        fields = [f for f in fields if not isinstance(f.function, (PETScArray, Temp))]
+        fields = [
+            f for f in fields if not (f.is_Dimension and not (f.is_Time or f.is_Modulo))
+        ]
+        return fields
+
+
+class ObjectBuilder:
+    """
+    A base class for constructing objects needed for a PETSc solver.
+    Designed to be extended by subclasses, which can override the `build`
+    method to support specific use cases.
```

Comment:
Some extended docstring explaining the content of. the dict below could be usefull